### PR TITLE
[Disk Manager] add vhd footer/header validations

### PR DESCRIPTION
**Pull request by BarkovBG**

### CODE REVIEW

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/footer.go [hunk 1]
```diff
@@ -39,5 +41,21 @@ func (f footer) validate() error {
 		)
 	}
 
-	return nil // TODO: Implement.
+	if f.FileFormatVersion != fileFormatVersion {
+		return common.NewSourceInvalidError(
+			"Failed to check vhd file format version: expected - %v, actual - %v",
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/footer.go:46]: We use format like that:
"expected %v, actual %v"

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/footer.go [hunk 2]
```diff
@@ -39,5 +41,21 @@ func (f footer) validate() error {
...
+			fileFormatVersion,
+			f.FileFormatVersion,
+		)
+	}
+
+	if f.DiskType != dynamicHardDiskType {
+		return common.NewSourceInvalidError(
+			"Failed to check vhd disk type: expected - %v, actual - %v",
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/footer.go:54]: Usually, error message is not capitalized

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/header.go [hunk 3]
```diff
@@ -6,26 +6,27 @@ import "github.com/ydb-platform/nbs/cloud/disk_manager/internal/pkg/dataplane/ur
 
 const (
 	// The dynamic disk header should appear on a sector (512-byte) boundary.
-	headerOffset = uint64(512)
-
-	headerCookie = "cxsparse"
+	headerOffset  = uint64(512)
+	headerCookie  = "cxsparse"
+	batOffset     = uint64(1536)
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/header.go:11]: What is "bat"?

Comment (BarkovBG) [reply cloud/disk_manager/internal/pkg/dataplane/url/vhd/header.go:11]: it is "block allocation table"

it was added here 
https://github.com/ydb-platform/nbs/commit/738a41e4b37e2a4d3c61a4eaa74accbc6fa781e0

Comment (SvartMetal) [reply cloud/disk_manager/internal/pkg/dataplane/url/vhd/header.go:11]: Need comment about that please

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/footer.go [hunk 4]
```diff
@@ -5,39 +5,57 @@ import "github.com/ydb-platform/nbs/cloud/disk_manager/internal/pkg/dataplane/ur
...
+	CreatorVersion     uint32    // This field holds the major/minor version of the application that created the hard disk image.
+	CreatorHostOS      uint32    // This field stores the type of host operating system this disk image is created on.
+	OriginalSize       uint64    // Stores the size of the hard disk in bytes at creation time.
+	CurrentSize        uint64    // Stores the current size of the hard disk in bytes.
+	DiskGeometry       uint32    // This field stores the cylinder, heads, and sectors per track value for the hard disk.
+	DiskType           uint32    // Type of the disk.
+	Checksum           uint32    // This field holds a basic checksum of the hard disk footer.
+	UniqueID           [16]byte  // А unique ID stored in the hard disk which is used to identify the hard disk.
+	SavedState         byte      // This field holds a one-byte flag that describes whether the system is in saved state.
+	Reserved           [427]byte // This field contains zeroes.
```
Comment (jkuradobery) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/footer.go:32]: The commend here is unnecessary.

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/footer.go [hunk 5]
```diff
@@ -5,39 +5,57 @@ import "github.com/ydb-platform/nbs/cloud/disk_manager/internal/pkg/dataplane/ur
...
-	return nil // TODO: Implement.
+	if f.FileFormatVersion != fileFormatVersion {
+		return common.NewSourceInvalidError(
+			"failed to check vhd file format version: expected %v, actual %v",
+			fileFormatVersion,
+			f.FileFormatVersion,
+		)
+	}
+
+	if f.DiskType != dynamicHardDiskType {
```
Comment (jkuradobery) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/footer.go:52]: Do you explicitly know that we would not parse fixed-sized vhd files? Also, the logic of VHD parsing for "Fixed hard disk" is missing. Maybe, add todo? I am open to discussion here.

**Review by jkuradobery — APPROVED** [review]

**Comment by jkuradobery** [comment]
Gofmt failed

**Review by jkuradobery — APPROVED** [review]

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/header.go [hunk 6]
```diff
@@ -6,36 +6,53 @@ import "github.com/ydb-platform/nbs/cloud/disk_manager/internal/pkg/dataplane/ur
...
+			"failed to check vhd header cookie: expected %s, actual %s",
 			headerCookie,
 			h.Cookie,
 		)
 	}
 
-	return nil // TODO: Implement.
+	if h.TableOffset != batOffset {
+		return common.NewSourceInvalidError(
+			"failed to check vhd header cookie: expected %v, actual %v",
```
Comment (gy2411) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/header.go:43]: cookie -> bat offset

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/footer.go [hunk 7]
```diff
@@ -5,39 +5,57 @@ import "github.com/ydb-platform/nbs/cloud/disk_manager/internal/pkg/dataplane/ur
...
-	DiskGeometry       uint32
-	DiskType           uint32
-	Checksum           uint32
-	UniqueID           [16]byte
-	SavedState         byte
+	Features           uint32   // A bit field used to indicate specific feature support.
+	FileFormatVersion  uint32   // Is divided into a major/minor version and matches the version of the specification used in creating the file.
+	DataOffset         uint64   // Absolute byte offset, from the beginning of the file, to the next structure.
+	Timestamp          uint32   // Creation time of a hard disk image.
+	CreatorApplication [4]byte  // Used to document which application created the hard disk.
```
Comment (gy2411) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/footer.go:22]: the hard disk image?

Comment (BarkovBG) [reply cloud/disk_manager/internal/pkg/dataplane/url/vhd/footer.go:22]: this comments are from official microsoft documentation

so, they use "hard disk" term. btw VHD is virtual HARD disk format

**Review by gy2411 — APPROVED** [review]

**Pull request merged**